src/why3prop.ml

```diff
@@ -0,0 +1,135 @@
+  ; symb_not   : sym (** Not(¬) symbol.                   *) }
+
+(** [get_prop_config pos builtins] set the builtins configuration using
+    [builtins] *)
+let get_prop_config :
+    Pos.popt -> sym StrMap.t -> prop_config = fun pos builtins ->
+    let find_sym key =
+        try StrMap.find key builtins with Not_found ->
+        Console.fatal pos "Builtin symbol [%s] undefined." key
+    in
+    { symb_P     = find_sym "P"
+    ; symb_T     = find_sym "T"
+    ; symb_or    = find_sym "or"
+    ; symb_and   = find_sym "and"
+    ; symb_imp   = find_sym "imp"
+    ; symb_bot   = find_sym "bot"
+    ; symb_top   = find_sym "top"
+    ; symb_not   = find_sym "not" }
+
+(** [translate pos builtins (hs, g)] translates from lambdapi to Why3 goal [g]
+    using the hypothesis [hs]. The function return
+    [constants_table, hypothesis, formula] where :
+    - [constants_table] maps abstracted Lambdapi terms to Why3 constants
+    - [hypothesis] maps abstracted labels to Why3 terms (presentation of [hs])
+    - [formula] Why3 term representing the goal [g].  *)
+
+let rec translate : Pos.popt -> sym StrMap.t -> (Env.env * term) ->
+    cnst_table * ctxt_labels * Why3.Term.term =
+    fun pos builtins  (hs, g) ->
+    let cfg = get_prop_config pos builtins in
+    let (constants_table, hypothesis) =
+        List.fold_left (translate_context cfg) ([], StrMap.empty) hs in
+    try
+        let (constants_table, formula) = t_goal cfg constants_table g in
+        (constants_table, hypothesis, formula)
+    with NoGoalTranslation ->
+        Console.fatal pos "The term [%a] is not of the form [P _]"
+        Print.pp g
+
+(** [translate_context cfg (l_constants, l_hypothesis) (hyp_name, (_, hyp))]
+    translate the context [hyp] with the label [hyp_name] and add it in
+    [l_hypothesis] with the why3 constants [l_constants]. *)
+and translate_context :
+    prop_config ->
+    cnst_table * ctxt_labels ->
+    string * (tvar * tbox) ->
+    cnst_table * ctxt_labels =
+    fun cfg (l_constants, l_hypothesis) (hyp_name, (_, hyp)) ->
+    try
+        let (new_why3_l, hyp') = t_goal cfg l_constants (Bindlib.unbox hyp) in
+            (new_why3_l, StrMap.add hyp_name hyp' l_hypothesis)
+    with NoGoalTranslation ->
+            (l_constants, l_hypothesis)
+
+(** [t_goal cfg constants_table trm] translate the lambdapi term [trm] to Why3
+    term using the configuration [cfg] and the list of Why3 constants in
+    [constants_table]. *)
+and t_goal : prop_config -> cnst_table -> term ->
+    cnst_table * Why3.Term.term =
+    fun cfg constants_table trm ->
+    match Basics.get_args trm with
+    | (symbol, [t]) when Basics.is_symb cfg.symb_P symbol ->
+        t_prop cfg constants_table [] t
+    | _                                                   ->
+        raise NoGoalTranslation
+
+(** [t_prop cfg constants_table ctxt p] translate the term [p] into Why3 terms
+    with a context [ctxt] and a config [cfg]. *)
+and t_prop :
```

Comment:
Thanks to rename t_prop into translate_prop.

Reply:
Done in a29414085332857be96b62ab67a96960b46d746d.